### PR TITLE
Support not exact matching of expected warnings

### LINKED ISSUE #50: [plugins] Support not exact matching of expected warnings
> need to add a property for EXACT match of warnings: exactWarningsMatch = true (default)
> What if save will expect 5 warnings and receive 6 (if static analyzer is not configurable and can show only all warnings by the default)

originally posted in https://github.com/cqfn/save/pull/39#discussion_r621986065

### PR DESCRIPTION
**Pull request by Cheshiriks**
### What's done:
* add exactWarningsMatch config
* Add warn test

Closes #50

### CODE REVIEW

File: save-core/src/commonMain/kotlin/org/cqfn/save/core/Save.kt [hunk 1]
```diff
@@ -47,8 +47,9 @@ class Save(
     @Suppress("WHEN_WITHOUT_ELSE")  // TestResult is a sealed class
     private fun handleResult(testResult: TestResult) {
         val status = testResult.status
+        val passMessage = "Test on resources [${testResult.resources}] has completed successfully"
```
Comment (kgevorkyan) [inline save-core/src/commonMain/kotlin/org/cqfn/save/core/Save.kt:50]: May be we can move initialization of this message into the block of Pass?
```kotlin
is Pass -> {
    val passMessage = "Test on resources [${testResult.resources}] has completed successfully"
    status.message?.let { logDebug("$passMessage. $it") } ?: logDebug(passMessage)
}
```

Comment (Cheshiriks) [reply save-core/src/commonMain/kotlin/org/cqfn/save/core/Save.kt:50]: Yes)

File: save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/WarnPlugin.kt [hunk 2]
```diff
@@ -75,31 +75,37 @@ class WarnPlugin : Plugin {
...
         val unexpectedWarnings = actualWarningsMap.filterValues { it !in expectedWarningsMap.values }.values
-        return when (missingWarnings.isEmpty() to unexpectedWarnings.isEmpty()) {
-            false to true -> "Some warnings were expected but not received: $missingWarnings"
-            false to false -> "Some warnings were expected but not received: $missingWarnings, " +
-                    "and others were unexpected: $unexpectedWarnings"
-            true to false -> "Some warnings were unexpected: $unexpectedWarnings"
-            true to true -> null
-            else -> ""
+        if (!warnPluginConfig.exactWarningsMatch && missingWarnings.isEmpty() && unexpectedWarnings.isNotEmpty()) {
+            return Pass("Some warnings were unexpected: $unexpectedWarnings")
```
[the remark this answers is not in the document]
Comment (Cheshiriks) [reply save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/WarnPlugin.kt]: Fixed it

File: save-plugins/warn-plugin/src/commonTest/kotlin/org/cqfn/save/plugin/warn/WarnPluginTest.kt [hunk 3]
```diff
@@ -265,4 +266,8 @@ class WarnPluginTest {
             )
         }
     }
+
+    companion object {
+        private const val NAME_WARN = "Some warnings were unexpected: [[Warning(message=Variable name should be in lowerCamelCase, line=5, column=8)]]"
```
Comment (petertrr) [inline save-plugins/warn-plugin/src/commonTest/kotlin/org/cqfn/save/plugin/warn/WarnPluginTest.kt:271]: But this is a text for a one particular test, why should it be a class-level constant?

File: save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/WarnPlugin.kt [hunk 4]
```diff
@@ -92,20 +92,17 @@ class WarnPlugin : Plugin {
                                        warnPluginConfig: WarnPluginConfig): TestStatus {
         val missingWarnings = expectedWarningsMap.filterValues { it !in actualWarningsMap.values }.values
         val unexpectedWarnings = actualWarningsMap.filterValues { it !in expectedWarningsMap.values }.values
-        if (!warnPluginConfig.exactWarningsMatch && missingWarnings.isEmpty() && unexpectedWarnings.isNotEmpty()) {
-            return Pass("Some warnings were unexpected: $unexpectedWarnings")
+
+        val isMissingWarnings = missingWarnings.isEmpty()
+        val isUnexpectedWarnings = unexpectedWarnings.isEmpty()
+        return when {
```
Comment (petertrr) [inline save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/WarnPlugin.kt:98]: Why did you remove the old version with `when (missingWarnings.isEmpty() to unexpectedWarnings.isEmpty())`?

**Review by petertrr — CHANGES_REQUESTED** [review]

File: save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/WarnPlugin.kt (hunk 4 above)
Comment (Cheshiriks) [reply save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/WarnPlugin.kt:98]: For one case, was needed a third parameter - warnPluginConfig.exactWarningsMatch

Comment (petertrr) [reply save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/WarnPlugin.kt:98]: You can place an additional `if` inside `when`'s branch. For me now it looks less clear, with multiple repeated variable names

Comment (Cheshiriks) [reply save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/WarnPlugin.kt]: Ok. Fixed it

File: save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/WarnPlugin.kt [hunk 5]
```diff
@@ -75,31 +75,35 @@ class WarnPlugin : Plugin {
...
-            false to true -> "Some warnings were expected but not received: $missingWarnings"
-            false to false -> "Some warnings were expected but not received: $missingWarnings, " +
-                    "and others were unexpected: $unexpectedWarnings"
-            true to false -> "Some warnings were unexpected: $unexpectedWarnings"
-            true to true -> null
-            else -> ""
+            false to true -> Fail("Some warnings were expected but not received: $missingWarnings")
+            false to false -> Fail("Some warnings were expected but not received: $missingWarnings, " +
+                    "and others were unexpected: $unexpectedWarnings")
+            true to false -> return if (!warnPluginConfig.exactWarningsMatch) {
```
Comment (petertrr) [inline save-plugins/warn-plugin/src/commonMain/kotlin/org/cqfn/save/plugin/warn/WarnPlugin.kt:100]:
```suggestion
            true to false -> if (!warnPluginConfig.exactWarningsMatch) {
```
`return` is redundant here?

**Review by petertrr — APPROVED** [review]
